src/Parser/GuardianDruid/Modules/Features/RageWasted.js

```diff
@@ -0,0 +1,132 @@
+import React from 'react';
+import { formatPercentage } from 'common/format';
+import SPELLS from 'common/SPELLS';
+import RESOURCE_TYPES from 'common/RESOURCE_TYPES';
+import SpellLink from 'common/SpellLink';
+import SpellIcon from 'common/SpellIcon';
+import StatisticBox from 'Main/StatisticBox';
+import Module from 'Parser/Core/Module';
+
+// NOTE: "Raw" rage is what shows up in combat log events (divided by 10 and rounded to get in-game rage).
+// We deal with raw rage here to prevent accuracy loss.
+
+// This is actually ~59.95, for some bizarre reason, meaning that occasionally you will gain 5 rage
+// instead of 6 from a melee.  We have no idea why this is. As long as we re-sync with the classResources
+// field with each event, this shouldn't be a problem.
+const RAW_RAGE_GAINED_FROM_MELEE = 60;
+
+// The cost of a Gory Fur-reduced Ironfur, or the minimum amount of rage that can be spent
+const MINIMUM_ACCEPTABLE_RAGE_WASTE = 34;
+
+const RAGE_GENERATORS = {
+  [SPELLS.MELEE.id]: 'Melee',
+  [SPELLS.MANGLE_BEAR.id]: 'Mangle',
+  [SPELLS.THRASH_BEAR.id]: 'Thrash',
+  [SPELLS.MOONFIRE.id]: 'Moonfire (Galactic Guardian)',
+  [SPELLS.BLOOD_FRENZY_TICK.id]: 'Blood Frenzy',
+  [SPELLS.BRISTLING_FUR.id]: 'Bristling Fur',
+  [SPELLS.OAKHEARTS_PUNY_QUODS_BUFF.id]: 'Oakheart\'s Puny Quods',
+  [SPELLS.PURE_RAGE_POTION.id]: 'Pure Rage Potion',
+};
+
+class RageWasted extends Module {
+  rageWastedBySpell = {};
+  totalRageGained = 0;
+  currentRawRage = 0;
+
+  // Currently always 1000, but in case a future tier set/talent/artifact trait increases this it should "just work"
+  currentMaxRage = 0;
+
+  synchronizeRage(event) {
+    if (!event.classResources) {
+      console.log('no classResources', event);
+      return;
+    }
+    const rageResource = event.classResources.find(resource => resource.type === RESOURCE_TYPES.RAGE);
+    if (rageResource) {
+      this.currentRawRage = rageResource.amount;
+      this.currentMaxRage = rageResource.max;
+    }
+  }
+
+  registerRageWaste(abilityID, waste) {
+    if (!this.rageWastedBySpell[abilityID]) {
+      this.rageWastedBySpell[abilityID] = waste;
+    } else {
+      this.rageWastedBySpell[abilityID] += waste;
+    }
+  }
+
+  on_byPlayer_energize(event) {
+    this.synchronizeRage(event);
+    if (event.resourceChangeType !== RESOURCE_TYPES.RAGE) {
+      return;
+    }
+
+    if (event.waste > 0) {
+      this.registerRageWaste(event.ability.guid, event.waste);
+      this.totalRageGained += event.resourceChange + event.waste;
+    }
+  }
+
+  on_byPlayer_cast(event) {
+    if (event.ability.guid === SPELLS.MELEE.id) {
+      if (this.currentRawRage + RAW_RAGE_GAINED_FROM_MELEE > this.currentMaxRage) {
+        const realRageWasted = Math.round((this.currentRawRage + RAW_RAGE_GAINED_FROM_MELEE - this.currentMaxRage) / 10);
+        this.registerRageWaste(event.ability.guid, realRageWasted);
```

Comment:
`event.resourceChange` is the real rage gain from the event. (Thrash generates 4 real rage)
`classResources.amount` is the current raw rage of the combatant after the event (41 real rage).

So the combatant was at 37 real rage (370 raw rage) before the event, then the event takes place and the combatant gains 4 real rage (40 raw rage) and now has 41 real rage (410 raw rage). 

When I handle melee attacks, I check that the combatant's raw rage before the melee attack plus the raw rage gain from melees does not exceed the maximum raw rage amount (I deal with raw rage because the values I have from `classResources` are already in raw rage and would be more accurate than converting them to real rage in the event that something like I described above occurs).  If it does exceed, I compute the wasted raw rage, convert it into real rage (by dividing by 10) and add it to the total wasted rage value (which is stored as real rage).